sdk/typescript/src/builder/TransactionBlock.ts

```diff
@@ -773,10 +774,15 @@ export class TransactionBlock {
 
 				if (initialSharedVersion) {
 					// There could be multiple transactions that reference the same shared object.
-					// If one of them is a mutable reference, then we should mark the input
+					// If one of them is a mutable reference or taken by value, then we should mark the input
 					// as mutable.
+					const isByValue =
+						normalizedType != null &&
+						extractMutableReference(normalizedType) == null &&
+						extractReference(normalizedType) == null;
 					const mutable =
 						isMutableSharedObjectInput(input.value) ||
+						isByValue ||
```

Comment:
A bit confused by this: what does it mean when a shared input object is passed here that's by value but without mutable flag? Or should we mandate by-value inputs to always have mutable flag on?